### PR TITLE
Routing

### PR DESCRIPTION
**Pull request by PavelBulat**

### CODE REVIEW

File: .env.default [hunk 1]
```diff
@@ -54,4 +54,7 @@ PAYMENT_SECRET_HEADER_NAME=X-Payment-Secret
 PAYMENT_SECRET_HEADER_VALUE=19pSIj1N5w5GFpeD88DssKEUe4QFwpE2
 
 #HTTP_AUTH_USER=
-#HTTP_AUTH_PASS=
\ No newline at end of file
+#HTTP_AUTH_PASS=
+
+STATIC_MAX_AGE=7d
+SETTINGS_NAME='falcon_default'
```
Comment (kalabro) [inline .env.default]: Can we rename `SETTINGS_NAME` to  `APPLICATION_NAME` and `STATIC_MAX_AGE` to `STATIC_CACHE_MAX_AGE`?

File: falcon/modules/features/falcon_feature_demo/config/install/menu_link_config.menu_link_config.create_a_new_article.yml [hunk 2]
```diff
@@ -5,7 +5,7 @@ id: create_a_new_article
 title: 'Create a new Article'
 route_name: node.add
 route_parameters:
-  node_type: news
+  node_type: 4b868b94-2d37-40f5-965a-f8922c65ba4c
```
Comment (kalabro) [inline falcon/modules/features/falcon_feature_demo/config/install/menu_link_config.menu_link_config.create_a_new_article.yml]: Can you check if it works? I think it should be `news`.

File: falcon/modules/features/falcon_feature_demo/config/install/menu_link_config.menu_link_config.create_a_new_page.yml [hunk 3]
```diff
@@ -5,7 +5,7 @@ id: create_a_new_page
 title: 'Create a new Page'
 route_name: node.add
 route_parameters:
-  node_type: page
+  node_type: a85c1597-ffcc-4912-9da8-497f85210f5a
```
Comment (kalabro) [inline falcon/modules/features/falcon_feature_demo/config/install/menu_link_config.menu_link_config.create_a_new_page.yml]: the same as with news link

File: falconjs/packages/falcon/package.json [hunk 4]
```diff
@@ -10,9 +10,17 @@
     "superagent-prefix": "^0.0.2"
   },
   "peerDependencies": {
-    "next": "^8.1.0",
+    "next": "^8.0.4",
     "react": "^16.8.6",
-    "react-dom": "^16.8.6"
+    "react-dom": "^16.8.6",
+    "redux": "^4.0.4",
+    "express-basic-auth": "^1.1.6",
```
Comment (kalabro) [inline falconjs/packages/falcon/package.json]: everything below `express-basic-auth` should be moved to dependencies.

File: falconjs/packages/falcon/routes/clearCache.js [hunk 5]
```diff
@@ -0,0 +1,17 @@
+const cache = require('memory-cache');
+const { parse } = require('url');
+
+function clearCache(req, res, next) {
+  // Get object with different parts of the URL.
+  const parsedUrl = parse(req.url, true);
+
+  // A little helper path to clear cached data.
+  if (parsedUrl.pathname === '/_cacheclear') {
```
Comment (kalabro) [inline falconjs/packages/falcon/routes/clearCache.js]: Move url to default/server.js (this middleware should not care about url).

File: falconjs/packages/falcon/routes/frontendOnlyRoutes.js [hunk 6]
```diff
@@ -0,0 +1,99 @@
+const pathToRegexp = require('path-to-regexp');
```
Comment (kalabro) [inline falconjs/packages/falcon/routes/frontendOnlyRoutes.js]: Rename routes/ to routing/ folder.

File: falconjs/packages/falcon/routes/getPageContent.js [hunk 7]
```diff
@@ -0,0 +1,89 @@
+const Router = require('next/router').default;
+const { parse } = require('url');
+const debug = require('debug')('falcon:routing/getPageContent');
+const { request: defaultRequest } = require('../request/request.node');
+const getEntityURL = require('../utils/getEntityURL');
+
+/**
+ * Returns Drupal entity data for both server side and client side.
+ */
+async function getPageContent(requestedURL, res = null, request = null) {
```
Comment (kalabro) [inline falconjs/packages/falcon/routes/getPageContent.js]: Rename to getEntityContent

File: falconjs/packages/falcon/routes/getPageContent.js [hunk 8]
```diff
@@ -0,0 +1,89 @@
...
+    entityURL: null,
+  };
+
+  const superagent = request || defaultRequest;
+
+  // Get object with different parts of the URL.
+  const parsedUrl = parse(requestedURL, true);
+
+  const entityURL = res && res.normalizedRequestPath
+    ? res.normalizedRequestPath : parsedUrl.pathname;
```
Comment (kalabro) [inline falconjs/packages/falcon/routes/getPageContent.js]: Move all custom res/req properties under falcon, i.e. `res.falcon.normalizedRequestPath`

File: falconjs/packages/falcon/routes/globalSettings.js [hunk 9]
```diff
@@ -0,0 +1,168 @@
...
+    props.statusCode = 200;
+  }
+
+  return props;
+}
+
+/**
+ * Returns the link object to the homepage.
+ */
+const getHomepageLink = (settings, isRaw = false) => {
```
Comment (kalabro) [inline falconjs/packages/falcon/routes/globalSettings.js]: Move away from server-side code (it looks like util or transform).

File: falconjs/packages/falcon/routes/globalSettings.js [hunk 10]
```diff
@@ -0,0 +1,168 @@
...
+        } else {
+          const { statusCode, settings } = await getSettings(settingsName, request, res);
+
+          // Handle Page Not Found response.
+          if (statusCode === 404) {
+            return await nextApp.render404(req, res, parsedUrl);
+          }
+
+          // Handle any other error.
+
```
Comment (kalabro) [inline falconjs/packages/falcon/routes/globalSettings.js]: why new line?

File: falconjs/packages/falcon/routes/globalSettings.js [hunk 11]
```diff
@@ -0,0 +1,168 @@
...
+      // At this point we already know that the page needs to be rendered either
+      // using internal routing of the app or using Drupal routing. So we can fetch
+      // global page settings (menu, header, footer) from backend or cache.
+      try {
+        // Use very simple in-memory cache storage for caching of global settings
+        // from Drupal. You can force flush it by restarting a server or
+        // just requesting /_cacheclear page if clearCache middleware is using.
+        const settingsFromCache = cache.get(settingsName);
+        if (settingsFromCache) {
+          res.settings = settingsFromCache;
```
Comment (kalabro) [inline falconjs/packages/falcon/routes/globalSettings.js]: res.falcon.settings

File: falconjs/packages/falcon/routes/globalSettings.js [hunk 12]
```diff
@@ -0,0 +1,168 @@
...
+        res.status(500);
+        return res.end('Internal Server Error');
+      }
+    }
+
+    return next();
+  };
+}
+
+function homePageInSettings(req, res, next) {
```
Comment (kalabro) [inline falconjs/packages/falcon/routes/globalSettings.js]: handleHomepageRequest

**Review by kalabro — CHANGES_REQUESTED** [review]

File: .env.default (hunk 1 above)
Comment (PavelBulat) [reply .env.default]: Done.

File: falcon/modules/features/falcon_feature_demo/config/install/menu_link_config.menu_link_config.create_a_new_article.yml (hunk 2 above)
Comment (PavelBulat) [reply falcon/modules/features/falcon_feature_demo/config/install/menu_link_config.menu_link_config.create_a_new_article.yml]: It doesn't work. Reverted.

File: falcon/modules/features/falcon_feature_demo/config/install/menu_link_config.menu_link_config.create_a_new_page.yml (hunk 3 above)
Comment (PavelBulat) [reply falcon/modules/features/falcon_feature_demo/config/install/menu_link_config.menu_link_config.create_a_new_page.yml]: Reverted.

File: falconjs/packages/falcon/package.json (hunk 4 above)
Comment (PavelBulat) [reply falconjs/packages/falcon/package.json]: Probably express-basic-auth should be moved to dependancies too. How do you think @kalabro ?

File: falconjs/packages/falcon/routes/clearCache.js (hunk 5 above)
Comment (PavelBulat) [reply falconjs/packages/falcon/routes/clearCache.js]: Done.

File: falconjs/packages/falcon/routes/frontendOnlyRoutes.js (hunk 6 above)
Comment (PavelBulat) [reply falconjs/packages/falcon/routes/frontendOnlyRoutes.js]: Done.

File: falconjs/packages/falcon/routes/getPageContent.js (hunk 7 above)
Comment (PavelBulat) [reply falconjs/packages/falcon/routes/getPageContent.js]: Done.

File: falconjs/packages/falcon/routes/getPageContent.js (hunk 8 above)
Comment (PavelBulat) [reply falconjs/packages/falcon/routes/getPageContent.js]: Done.

File: falconjs/packages/falcon/routes/globalSettings.js (hunk 9 above)
Comment (PavelBulat) [reply falconjs/packages/falcon/routes/globalSettings.js]: Moved to utils as new file.

File: falconjs/packages/falcon/routes/globalSettings.js (hunk 10 above)
Comment (PavelBulat) [reply falconjs/packages/falcon/routes/globalSettings.js]: Removed.

File: falconjs/packages/falcon/routes/globalSettings.js (hunk 11 above)
Comment (PavelBulat) [reply falconjs/packages/falcon/routes/globalSettings.js]: Done.

File: falconjs/packages/falcon/routes/globalSettings.js (hunk 12 above)
Comment (PavelBulat) [reply falconjs/packages/falcon/routes/globalSettings.js]: Done.

File: falconjs/packages/falcon/routes/frontendOnlyRoutes.js [hunk 13]
```diff
@@ -0,0 +1,99 @@
...
+      value = value.join('/');
+    }
+    return [
+      encodeURIComponent(key),
+      encodeURIComponent(value),
+    ].join('=');
+  }).join('&');
+
+
+function frontendOnlyRoutes(nextApp, routes) {
```
Comment (kitnato) [inline falconjs/packages/falcon/routes/frontendOnlyRoutes.js]: We seem to have a mix of named functions with ES6 function syntax (e.g. line 61) alongside these anonymous declarations in several files. Wouldn't it better if we made all function declarations consistent with ES6 syntax in the `const` style?

Comment (kitnato) [reply falconjs/packages/falcon/routes/frontendOnlyRoutes.js]: An example with `async`: 
```js
const funcName = async (param) => {};
```

**Pull request merged**